### PR TITLE
Fix a delimiter bug if external table has delimiter 'OFF', and the value of the column is 'O'.

### PR DESCRIPTION
**Pull request by linwen**
Add a bool flag 'delim_off' for CopyStateData to indicate if delimiter is set to OFF or not.

This PR is to fix: https://github.com/greenplum-db/gpdb/issues/10009

### CODE REVIEW

File: src/backend/commands/copy.c [hunk 1]
```diff
@@ -6594,7 +6594,7 @@ CopyReadAttributesText(CopyState cstate, int stop_processing_at_field)
 			if (cur_ptr >= line_end_ptr)
 				break;
 			c = *cur_ptr++;
-			if (c == delimc)
+			if (c == delimc && !cstate->delim_off)
```
Comment (hlinnaka) [inline src/backend/commands/copy.c:6597]: Please note that this loop is *very* performance sensitive. Even small overheads here matter. The 'delimc' is copied to a local variable here, for performance reasons I believe. Should we do the same for delim_off?

If we set delimc = '\0' when delim_off is true, can we skip the check for 'delim_off' here altogether? I don't think a nul byte can appear in the input.

Comment (linwen) [reply src/backend/commands/copy.c]: Heikki, thanks for your review comments! I have use a local variable to replace pointer access. 
"If we set delimc = '\0' when delim_off is true, can we skip the check for 'delim_off' here altogether?" about this, I've asked @lij55 , it is possible to have '\0' in customer's input data.

Comment (lij55) [reply src/backend/commands/copy.c]: thanks for the review, @hlinnaka .  The most common use case for `delimiter off` is to handle multiple-byte delimiter or special character delimiter (including \0).

**Comment by pf-qiu** [comment]
I think we also need to make the same performance fix to master

**Comment by linwen** [comment]
> I think we also need to make the same performance fix to master

Thanks!  I will fix it too for master branch.

**Pull request merged**